packages/flutter_tools/lib/src/resident_runner.dart

```diff
@@ -254,26 +261,29 @@ class FlutterDevice {
       await device.stopApp(package);
       return;
     }
-    final List<FlutterView> flutterViews = views;
-    if (flutterViews == null || flutterViews.isEmpty) {
+    await refreshViews();
+    if (views == null || views.isEmpty) {
       return;
     }
     // If any of the flutter views are paused, we might not be able to
     // cleanly exit since the service extension may not have been registered.
-    if (flutterViews.any((FlutterView view) {
-      return view != null &&
-             view.uiIsolate != null &&
-             view.uiIsolate.pauseEvent != null &&
-             view.uiIsolate.pauseEvent.isPauseEvent;
+    for (final FlutterView flutterView in views) {
+      final vm_service.Isolate isolate = await vmService
+        .getIsolate(flutterView.uiIsolate.id)
+        .catchError((dynamic error, StackTrace stackTrace) {
+          return null;
+        }, test: (dynamic error) => error is vm_service.SentinelException);
+      if (isolate == null) {
+        continue;
+      }
+      if (isPauseEvent(isolate.pauseEvent.kind)) {
+        await device.stopApp(package);
+        return;
       }
-    )) {
-      await device.stopApp(package);
-      return;
     }
     final List<Future<void>> futures = <Future<void>>[];
-    for (final FlutterView view in flutterViews) {
+    for (final FlutterView view in views) {
       if (view != null && view.uiIsolate != null) {
-        assert(!view.uiIsolate.pauseEvent.isPauseEvent);
         futures.add(vmService.flutterExit(
           isolateId: view.uiIsolate.id,
         ));
```

Comment:
Regarding the comment below on 282, what would go wrong if we don't wait for `flutterExit` to fail?

Reply:
Its not if it fails its that when it succeeds the engine goes down so there is no signal. I could re-arrange this to instead listen to the vm_service `done` signal, that would remove the need for the timeout

Reply:
Updated the logic to be a bit simpler, and catch any SentinelException (would be thrown if we exited after the app crashed, for example)